File: Semester 4/Medii de proiectare si programare/Laborator/Laborator 6/FestivalMuzicaJavaFXClientServer/services/src/main/java/ro/mpp2026/festivalmuzicajavafx/service/ServiceImpl.java
package ro.mpp2026.festivalmuzicajavafx.service;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2026.festivalmuzicajavafx.domain.Show;
import ro.mpp2026.festivalmuzicajavafx.domain.ShowFilter;
import ro.mpp2026.festivalmuzicajavafx.domain.Ticket;
import ro.mpp2026.festivalmuzicajavafx.domain.User;
import ro.mpp2026.festivalmuzicajavafx.repository.ShowsRepository;
import ro.mpp2026.festivalmuzicajavafx.repository.TicketsRepository;
import ro.mpp2026.festivalmuzicajavafx.repository.UsersRepository;
import ro.mpp2026.festivalmuzicajavafx.utils.Observer;

import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;
import java.time.LocalDate;
import java.util.*;
import java.util.stream.StreamSupport;

public class ServiceImpl implements Service {
    private final UsersRepository repository;
    private final String encryptionAESKey;
    private final Map<String, Observer> loggedClients;
    private final ShowsRepository showsRepository;
    private final TicketsRepository ticketsRepository;
    private static final Logger logger = LogManager.getLogger(ServiceImpl.class);

    public ServiceImpl(UsersRepository usersRepository, ShowsRepository showsRepository, TicketsRepository ticketsRepository, String encryptionAESKey) {
        this.repository = usersRepository;
        this.encryptionAESKey = encryptionAESKey;
        this.loggedClients = new HashMap<>();
        this.showsRepository = showsRepository;
        this.ticketsRepository = ticketsRepository;
    }

    public User login(String email, String password, Observer client) {
        String encryptedPassword = encrypt(password, encryptionAESKey);

        Optional<User> user = repository.findByEmailAndPassword(email, encryptedPassword);

        if(user.isEmpty())
            throw new RuntimeException("Invalid credentials");

        if(loggedClients.containsKey(user.get().getId().toString()))
            throw new RuntimeException("User already logged in");

        loggedClients.put(user.get().getId().toString(), client);
        return user.get();
    }

    public void register(String email, String password) {
        String encryptedPassword = encrypt(password, encryptionAESKey);

        Optional<User> user = repository.findByEmailAndPassword(email, encryptedPassword);

        if(user.isPresent())
            throw new RuntimeException("Email already exists");

        repository.save(new User(0L, email, encryptedPassword));
    }

    @Override
    public void logout(Long userId, Observer client) {
        loggedClients.remove(userId.toString());
    }

    private String encrypt(String data, String key) {
        SecretKeySpec secretKey = new SecretKeySpec(key.getBytes(), "AES");

        Cipher cipher = null;
        try {
            cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, secretKey);

            byte[] encryptedBytes = cipher.doFinal(data.getBytes());
            return Base64.getEncoder().encodeToString(encryptedBytes);
        }
        catch (Exception exception) {
            throw new RuntimeException("Password encryption failed!");
        }
    }

    public void save(String artistName, LocalDate date, String location, int remainingSeats) {
        Show show = new Show(0L, artistName, date, location, remainingSeats);
        showsRepository.save(show);
        notifyObservers();
    }

    public List<Show> findAll() {
        return StreamSupport.stream(showsRepository.findAll().spliterator(), false)
                .toList();
    }

    public List<Show> findAll(ShowFilter showFilter) {
        return StreamSupport.stream(showsRepository.findAll(showFilter).spliterator(), false)
                .toList();
    }

    public Optional<Show> findById(Long id) {
        Show show = showsRepository.findOne(id);
        return Optional.ofNullable(show);
    }

    public void bookTicketForShow(Long showId, String client, Integer numberOfSeats) {
        Show show = showsRepository.findOne(showId);

        if(show.getRemainingSeats() < numberOfSeats)
            throw new RuntimeException("Not enough seats");

        Ticket ticket = new Ticket(0L, client, show, numberOfSeats);
        show.setRemainingSeats(show.getRemainingSeats() - numberOfSeats);

        ticketsRepository.save(ticket);
        showsRepository.update(showId, show);

        notifyObservers();
    }

    public void updateTicket(Long ticketId, String client, Long showId, Integer numberOfSeats) {
        Ticket ticket = ticketsRepository.findOne(ticketId);
        Show show = ticket.getShow();

        if(show.getRemainingSeats() + ticket.getNoSeats() - numberOfSeats < 0)
            throw new RuntimeException("Not enough seats");

        show.setRemainingSeats(show.getRemainingSeats() + ticket.getNoSeats() - numberOfSeats);
        ticket.setNoSeats(numberOfSeats);

        ticketsRepository.update(ticketId, ticket);
        showsRepository.update(show.getId(), show);

        notifyObservers();
    }

    public List<Ticket> findAllTickets() {
        return StreamSupport.stream(ticketsRepository.findAll().spliterator(), false)
                .toList();
    }

    public List<Ticket> findAllTicketsForShow(Long showId) {
        return StreamSupport.stream(ticketsRepository.findAllForShow(showId).spliterator(), false)
                .toList();
    }

    public int getNumberOfSoldSeatsForShow(Long showId) {
        return findAllTicketsForShow(showId).stream()
                .mapToInt(Ticket::getNoSeats)
                .sum();
    }

    @Override
    public void subscribe(Observer observer) {
    }

    @Override
    public void unsubscribe(Observer observer) {  }

    @Override
    public void notifyObservers() {
        for(var observer : loggedClients.values())
            observer.update();
    }
}

File: Semester 4/Medii de proiectare si programare/Laborator/Laborator 10/FestivalMuzicaCSharpWFA/server/ServiceImpl.cs
using FestivalMuzicaCSharp.Domain;
using FestivalMuzicaCSharp.Repository;
using log4net;
using services.Utils;
using System.Configuration;
using System.Security.Cryptography;
using System.Text;


namespace server;
public class ServiceImpl : IService
{
    private IUsersRepository usersRepository;
    private IShowsRepository showsRepository;
    private ITicketsRepository ticketsRepository;
    private readonly IDictionary<String, IObserver> loggedClients;
    private static readonly ILog log = LogManager.GetLogger(typeof(ServiceImpl));
    private readonly string _encryptionAESKey;

    public ServiceImpl(IUsersRepository usersRepository, IShowsRepository showsRepository, ITicketsRepository ticketsRepository)
    {
        this.usersRepository = usersRepository;
        this.showsRepository = showsRepository;
        this.ticketsRepository = ticketsRepository;
        loggedClients = new Dictionary<String, IObserver>();
        _encryptionAESKey = ConfigurationManager.AppSettings["encryptionAESKey"];
        if (string.IsNullOrEmpty(_encryptionAESKey))
            throw new Exception("Missing encryptionAESKey in app.config <appSettings>");
    }

    public void BookTicketForShow(long showId, string clientName, int noSeats)
    {
        var show = showsRepository.FindOne(showId);
        if (show == null)
            throw new Exception("Show not found");

        if (show.RemainingSeats < noSeats)
            throw new Exception("Not enough seats available");

        var ticket = new Ticket(0, clientName, show, noSeats);
        show.RemainingSeats -= noSeats;

        ticketsRepository.Save(ticket);
        showsRepository.Update(showId, show);

        NotifyObservers();
    }

    public List<Show> FindAll()
    {
        return [.. showsRepository.FindAll()];
    }

    public List<Show> FindAll(ShowFilter showFilter)
    {
        List<Show> shows = [.. showsRepository.FindAll()];

        if (showFilter?.ArtistName != null)
            shows = shows.Where(s => s.ArtistName.Contains(showFilter.ArtistName)).ToList();

        if (showFilter?.Date != null)
            shows = shows.Where(s => s.Date.Equals(showFilter.Date)).ToList();

        if (showFilter?.Location != null)
            shows = shows.Where(s => s.Location.Contains(showFilter.Location)).ToList();

        if (showFilter?.RemainingSeats != null)
            shows = shows.Where(s => s.RemainingSeats.Equals(showFilter.RemainingSeats)).ToList();

        return shows;
    }

    public List<Ticket> FindAllTickets()
    {
        return [.. ticketsRepository.FindAll()];
    }

    public List<Ticket> FindAllTicketsForShow(long showId)
    {
        return [.. ticketsRepository.FindAll().Where(t => t.Show.Id == showId)];
    }

    public Show? FindOne(long id)
    {
        return showsRepository.FindOne(id);
    }

    public int GetNumberOfSoldSeatsForShow(long showId)
    {
        return ticketsRepository.FindAll().Where(t => t.Show.Id == showId).Sum(t => t.NoSeats);
    }

    public void Save(string artistName, DateOnly date, string location, int noSeats)
    {
        var show = new Show(0, artistName, date, location, noSeats);
        showsRepository.Save(show);
        NotifyObservers();
    }

    public void UpdateTicket(long ticketId, long showId, string clientName, int noSeats)
    {
        var ticket = ticketsRepository.FindOne(ticketId) ?? throw new Exception("Ticket not found");

        var show = ticket.Show;
        int seatDifference = noSeats - ticket.NoSeats;

        if (show.RemainingSeats < seatDifference)
            throw new Exception("Not enough seats available");

        show.RemainingSeats -= seatDifference;
        ticket.ClientName = clientName;
        ticket.NoSeats = noSeats;

        ticketsRepository.Update(ticketId, ticket);
        showsRepository.Update(show.Id, show);

        NotifyObservers();
    }

    public void NotifyObservers()
    {
        //Console.WriteLine("~~~~~~~~Notific");
        foreach (var item in loggedClients)
        {
            Task.Run(() =>
            {
                //Console.WriteLine("item");
                item.Value.Update();
            });
        }
    }

    public User Login(string email, string password, IObserver client)
    {
        string encryptedPassword = Encrypt(password, _encryptionAESKey);

        var user = usersRepository.FindByEmailAndPassword(email, encryptedPassword);
        if (user != null)
        {
            if (loggedClients.ContainsKey(user.Id.ToString()))
                throw new Exception("User already logged in");
            loggedClients[user.Id.ToString()] = client;
            return user;
        }

        throw new Exception("Invalid credentials");
    }

    public void Logout(long userId, IObserver client)
    {
        _ = loggedClients[userId.ToString()] ?? throw new Exception("User " + userId + " is not logged in.");
        loggedClients.Remove(userId.ToString());
    }

    public void Register(string email, string password)
    {
        string encryptedPassword = Encrypt(password, _encryptionAESKey);

        var existingUser = usersRepository.FindByEmail(email);
        if (existingUser != null)
            throw new Exception("User already exists with this email");

        usersRepository.Save(new User(0, email, encryptedPassword));
    }

    public string Encrypt(string data, string key)
    {
        try
        {
            using Aes aesAlg = Aes.Create();
            aesAlg.Key = Encoding.UTF8.GetBytes(key.PadRight(32).Substring(0, 32));
            aesAlg.Mode = CipherMode.ECB;
            aesAlg.Padding = PaddingMode.PKCS7;
            using ICryptoTransform encryptor = aesAlg.CreateEncryptor();
            byte[] inputBytes = Encoding.UTF8.GetBytes(data);
            byte[] encrypted = encryptor.TransformFinalBlock(inputBytes, 0, inputBytes.Length);
            return Convert.ToBase64String(encrypted);
        }
        catch
        {
            throw new Exception("Password encryption failed!");
        }
    }

    public void Subscribe(IObserver observer)
    {
        throw new NotImplementedException();
    }

    public void Unsubscribe(IObserver observer)
    {
        throw new NotImplementedException();
    }
}
